framework/src/variables/MooseVariableDataBase.C

```diff
@@ -0,0 +1,827 @@
+    return _vector_tag_u[tag];
+  else
+    mooseError("Tag ", tag, " is not associated with any vector for variable ", _var.name());
+}
+
+template <typename OutputType>
+const typename MooseVariableDataBase<OutputType>::DoFValue &
+MooseVariableDataBase<OutputType>::vectorTagDofValue(TagID tag) const
+{
+  if (tag >= _need_vector_tag_dof_u.size())
+    const_cast<MooseVariableDataBase<OutputType> *>(this)->resizeVectorTagData(tag);
+
+  _need_vector_tag_dof_u[tag] = true;
+
+  if (_sys.hasVector(tag))
+    return _vector_tags_dof_u[tag];
+  else
+    mooseError("Tag ", tag, " is not associated with any vector for variable ", _var.name());
+}
+
+template <typename OutputType>
+const typename MooseVariableDataBase<OutputType>::FieldVariableGradient &
+MooseVariableDataBase<OutputType>::vectorTagGradient(TagID tag) const
+{
+  if (tag >= _need_vector_tag_grad.size())
+    const_cast<MooseVariableDataBase<OutputType> *>(this)->resizeVectorTagData(tag);
+
+  _need_vector_tag_grad[tag] = true;
+
+  if (_sys.hasVector(tag))
+    return _vector_tag_grad[tag];
+  else
+    mooseError("Tag ", tag, " is not associated with any vector for variable ", _var.name());
+}
+
+template <typename OutputType>
+const typename MooseVariableDataBase<OutputType>::FieldVariableValue &
+MooseVariableDataBase<OutputType>::matrixTagValue(TagID tag) const
+{
+  if (tag >= _matrix_tag_u.size())
+  {
+    _need_matrix_tag_u.resize(tag + 1, false);
+    const_cast<MooseVariableDataBase<OutputType> *>(this)->_matrix_tag_u.resize(tag + 1);
+  }
+
+  _need_matrix_tag_u[tag] = true;
+
+  if (_sys.hasMatrix(tag))
+    return _matrix_tag_u[tag];
+  else
+    mooseError("Tag ", tag, " is not associated with any matrix for variable ", _var.name());
+}
+
+template <typename OutputType>
+unsigned int
+MooseVariableDataBase<OutputType>::oldestSolutionStateRequested() const
+{
+  return _max_state;
+}
+
+template <typename OutputType>
+void
+MooseVariableDataBase<OutputType>::needSolutionState(const unsigned int state)
+{
+  if (state > _max_state)
+  {
+    _sys.needSolutionState(state);
+    _max_state = state;
+  }
+}
+
+template <typename OutputType>
+template <typename ReturnType, typename Functor>
+const ReturnType &
+MooseVariableDataBase<OutputType>::stateToTagHelper(const Moose::SolutionState state,
+                                                    Functor functor)
+{
+  if (state > 0)
+    // We need to request all states that are between current and the requested state
+    stateToTagHelper<ReturnType>(Moose::SolutionState(static_cast<int>(state) - 1), functor);
+
+  switch (state)
+  {
+    case Moose::Current:
+      return functor(_subproblem.getVectorTagID(Moose::SOLUTION_TAG));
+
+    case Moose::Old:
+    {
+      needSolutionState(1);
+      _old_solution_tag = _subproblem.getVectorTagID(Moose::OLD_SOLUTION_TAG);
+      insertSolutionTag(_old_solution_tag);
+      return functor(_old_solution_tag);
+    }
+
+    case Moose::Older:
+    {
+      needSolutionState(2);
```

Comment:
Use enum would be better

Reply:
I like the integral argument for as many methods as possible. At some point we may want a solution state 3, and then adding an Olderer enum option would feel a little silly. I can remove these lines though